File: leetcodely/javascript/ds/singlyLinkedList.js
class SinglyLinkedList{

    constructor(){
        this.head = null;
        this.size = 0;
    }

    insert(item){
        this.size ++;
        let p = new Node(item);
        if(this.head === null){
            this.head = p;
        }
        else{
           p.next = this.head;
           this.head = p;
        }
    }
    insertAt(item, index){
        if(index > this.size){
            throw "Wrong index";
        }
        let p = new Node(item);
        if(index === 0){
            p.next = this.head;
            this.head = p;
            return;
        }
        let curr = this.head;
        while(curr !== null){
            if(index === 0){
                if(curr.next === null){
                    curr.next = p;
                    return;
                }
                else if(curr.next.next === null){
                    curr.next = p;
                }else{
                    curr.next.next = p.next;
                    curr.next = p;

                }
                return;
            }
            curr = curr.next;
            index --;
        }

    }
    remove(item){
        let curr, prev;
        prev = curr = this.head;
        if(item === curr.val){
            this.head = this.head.next;
            this.size --;
            return;
        }
        while(curr !== null){
            if (curr.val === item){
                prev.next = curr.next;
                this.size --;
                return;
            }
            prev = curr;
            curr = curr.next;

        }
        throw "Item doesn't exist in list."
    }
    length(){
        return this.size;
    }

    isEmpty(){
        return this.size === 0;
    }
    stringify(){
        let curr = this.head;
        const out = [];
        while(curr !== null){
            if(curr.next === null){
                out.push(curr.val);
            }else{
                out.push(curr.val);
                out.push("->")
            }
            curr = curr.next;
        }
        return out.join("");
    }
}

class Node{
    constructor(val){
        this.val = val;
        this.next = null;
    }
}

module.exports = SinglyLinkedList;

File: leetcodely/go/ds/singlyLinkedList.go
package ds

import (
	"errors"
	"fmt"
)

type node struct {
	data int
	next *node
}

type SinglyLinkedList struct {
	head *node
	size int
}

func NewSinglyLinkedList() *SinglyLinkedList {
	return &SinglyLinkedList{
		head: nil,
	}
}

func (l *SinglyLinkedList) Insert(data int) error {
	n := &node{
		data: data,
	}
	if l.head == nil {
		l.head = n
	} else {
		n.next = l.head
		l.head = n
	}
	l.size++
	return nil
}

func (l *SinglyLinkedList) Remove(data int) error {
	if l.head.data == data {
		l.head = l.head.next

	} else {
		curr := l.head
		for !(curr == nil) {
			if curr.next == nil {
				return errors.New("item doesn't exist in the list")
			} else if curr.next.data == data {
				curr.next = curr.next.next
			}
			curr = curr.next
		}
	}
	return nil
}

func (l *SinglyLinkedList) Print() error {
	curr := l.head
	if curr == nil {
		return errors.New("Empty List")
	}
	for curr != nil {
		fmt.Println(curr.data)
		curr = curr.next
	}

	return nil
}
